File: packages/urdu-js/src/encoding/inpage/detect.ts
import type { Encoding } from '../../types.js'

export function detectEncoding(buffer: Uint8Array): Encoding {
  if (buffer.length >= 2 && buffer[0] === 0xff && buffer[1] === 0xfe) return 'utf-16le'
  if (buffer.length >= 2 && buffer[0] === 0xfe && buffer[1] === 0xff) return 'utf-16be'
  if (buffer.length >= 3 && buffer[0] === 0xef && buffer[1] === 0xbb && buffer[2] === 0xbf) return 'utf-8'
  const sample = buffer.slice(0, Math.min(512, buffer.length))
  let inpageCount = 0
  for (const b of sample) if (b === 0x04) inpageCount++
  if (inpageCount / Math.max(sample.length, 1) > 0.05) return 'inpage-v1v2'
  let urduCount = 0
  for (let i = 0; i + 1 < sample.length; i += 2) if (sample[i + 1] === 0x06) urduCount++
  if (urduCount / Math.max(sample.length / 2, 1) > 0.1) return 'inpage-v3'
  return 'utf-8'
}

File: packages/urdu-dart/lib/src/encoding/inpage/detect.dart
import 'dart:math';
import 'dart:typed_data';
import '../../types.dart';

/// Detects the encoding of a given binary buffer.
Encoding detectEncoding(Uint8List buffer) {
  if (buffer.length >= 2 && buffer[0] == 0xff && buffer[1] == 0xfe)
    return Encoding.utf16le;
  if (buffer.length >= 2 && buffer[0] == 0xfe && buffer[1] == 0xff)
    return Encoding.utf16be;
  if (buffer.length >= 3 &&
      buffer[0] == 0xef &&
      buffer[1] == 0xbb &&
      buffer[2] == 0xbf) return Encoding.utf8;

  final sampleLen = min(512, buffer.length);
  final sample = buffer.sublist(0, sampleLen);

  int inpageCount = 0;
  for (final b in sample) {
    if (b == 0x04) inpageCount++;
  }
  if (inpageCount / max(sample.length, 1) > 0.05) return Encoding.inpageV1v2;

  int urduCount = 0;
  for (var i = 0; i + 1 < sample.length; i += 2) {
    if (sample[i + 1] == 0x06) urduCount++;
  }
  if (urduCount / max(sample.length / 2, 1) > 0.1) return Encoding.inpageV3;

  return Encoding.utf8;
}
